webapp/src/main/java/com/box/l10n/mojito/android/strings/AndroidStringDocumentMapper.java

```diff
@@ -0,0 +1,186 @@
+                                       String assetDelimiter,
+                                       String locale,
+                                       String repositoryName) {
+        this.pluralSeparator = pluralSeparator;
+        this.assetDelimiter = Optional.ofNullable(Strings.emptyToNull(assetDelimiter)).orElse(DEFAULT_ASSET_DELIMITER);
+        this.locale = locale;
+        this.repositoryName = repositoryName;
+    }
+
+    public AndroidStringDocumentMapper(String pluralSeparator, String assetDelimiter) {
+        this(pluralSeparator, assetDelimiter, null, null);
+    }
+
+    public AndroidStringDocument readFromTextUnits(List<TextUnitDTO> textUnits, boolean useSource) {
+
+        AndroidStringDocument document = new AndroidStringDocument();
+        Map<String, AndroidPlural.AndroidPluralBuilder> pluralByOther = new HashMap<>();
+
+        textUnits.forEach(textUnit -> {
+
+            if (isSingularTextUnit(textUnit)) {
+                document.addSingular(textUnitToAndroidSingular(textUnit, useSource));
+            } else {
+                pluralByOther.compute(textUnit.getPluralFormOther(), (key, builder) -> {
+                    if (Objects.isNull(builder)) {
+                        builder = AndroidPlural.builder();
+                    }
+
+                    if (textUnit.getPluralForm().equalsIgnoreCase(OTHER.name())) {
+                        String name = PluralNameParser.getPrefix(textUnit.getName(), pluralSeparator);
+                        builder.setName(textUnit.getAssetPath() + assetDelimiter + name);
+                        builder.setComment(textUnit.getComment());
+                    }
+
+                    builder.addItem(new AndroidPluralItem(
+                            textUnit.getPluralForm(),
+                            textUnit.getTmTextUnitId(),
+                            useSource ? textUnit.getSource() : textUnit.getTarget()));
+
+                    return builder;
+
+                });
+            }
+        });
+
+        pluralByOther.forEach((pluralFormOther, builder) -> document.addPlural(builder.build()));
+
+        return document;
+    }
+
+    public AndroidStringDocument readFromSourceTextUnits(List<TextUnitDTO> textUnits) {
+        return readFromTextUnits(textUnits, true);
+    }
+
+    public AndroidStringDocument readFromTargetTextUnits(List<TextUnitDTO> textUnits) {
+        return readFromTextUnits(textUnits, false);
+    }
+
+    public List<TextUnitDTO> mapToTextUnits(AndroidStringDocument document) {
+        return mapToTextUnitStream(document).collect(Collectors.toList());
+    }
+
+    public Stream<TextUnitDTO> mapToTextUnitStream(AndroidStringDocument document) {
+        return document.getStrings()
+                .stream()
+                .flatMap(this::stringToTextUnits)
+                .map(this::addTextUnitDTOAttributes);
+    }
+
+    TextUnitDTO addTextUnitDTOAttributes(TextUnitDTO textUnit) {
```

Comment:
why isn't this done in "singularToTextUnit" and "pluralToTextUnits"

Reply:
We need to apply this method to all text units, regardless of them being singular or plural.

Reply:
yes, my point was the conversion functions "singularToTextUnit" and "pluralToTextUnits" should be self sufficient and not expect other logic to run later to complete the objects

Reply:
Updated.